README.md
# django-wizard-builder

| Status | Support |
| --- | --- |
| [![Build Status][build-img]][build-url] | ![python 3.6][python36] |
| [![PyPI Version][pypi-img]][pypi-url] | ![django 1.11][django111] |

[build-img]: https://travis-ci.org/SexualHealthInnovations/django-wizard-builder.png?branch=master
[build-url]: https://travis-ci.org/SexualHealthInnovations/django-wizard-builder
[pypi-img]: https://img.shields.io/pypi/v/django-wizard-builder.svg
[pypi-url]: https://pypi.python.org/pypi/django-wizard-builder

[python36]: https://img.shields.io/badge/python-3.6-green.svg
[django111]: https://img.shields.io/badge/django-1.11-yellowgreen.svg


## Installation

django settings

    INSTALLED_APPS = [
        'django.contrib.sites',
        'nested_admin',
        'widget_tweaks',
        'wizard_builder',
    ]


urls.py

    urlpatterns = [
        url(r'^nested_admin/', include('nested_admin.urls')),
        url(r'^admin/', admin.site.urls),
    ]


## Developement

Welcome to the Callisto Product Team! Here's a list of things you'll need to get started:

    $ make install-osx  # on osx
    $ make app-setup


cleanup / linting / tests

    $ make clean-lint
    $ make test-fast


HISTORY.md
# History / Changelog

## 2.5.7 (2017-11-23)

* maintainence release

## 2.5.6 (2017-11-22)

* supress extra question inlines

## 2.5.5 (2017-11-21)

* add a missing migration

## 2.5.4 (2017-11-21)

* remove direct use of get_current_site

## 2.5.3 (2017-11-06)

* all foreign keys casade deletion now

## 2.5.2 (2017-11-02)

* re-enable conditionals

    https://github.com/project-callisto/django-wizard-builder/pull/205

## 2.5.1 (2017-10-30)

* fix migration 0034 on postgres

## 2.5.0 (2017-10-28)

* removed downcasting queries
* removed django-model-utils dependency
* refactored admin classes

    https://github.com/project-callisto/django-wizard-builder/pull/200

## 2.4.4 (2017-10-28)

* fix button order in example app

## 2.4.3 (2017-10-20)

* partially disable conditionals

    https://github.com/project-callisto/django-wizard-builder/pull/198

## 2.4.2 (2017-10-19)

* answer display updates for [callisto-core data persistence PR](https://github.com/project-callisto/callisto-core/pull/311)

    https://github.com/project-callisto/django-wizard-builder/pull/197

## 2.4.1 (2017-10-17)

* move field generating code out of the models

    https://github.com/project-callisto/django-wizard-builder/pull/196

## 2.4.0 (2017-10-17)

* question fields are now generated from session data, rather than being pulled from the database on each request

    https://github.com/project-callisto/django-wizard-builder/pull/192

## 2.3.0 (2017-10-12)

* add TextArea question type (s/o [@scattermagic](https://github.com/scattermagic))

    https://github.com/project-callisto/django-wizard-builder/pull/191

## 2.2.4 (2017-10-12)

* fix multiple choice answer display on review page (s/o [@scattermagic](https://github.com/scattermagic))

    https://github.com/project-callisto/django-wizard-builder/pull/187

## 2.2.3 (2017-09-19)

* fix a bug with incorrectly formatted review page answers

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/177

## 2.2.2 (2017-09-18)

* update tests for callisto's use case
* remove infobox

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/176

## 2.2.1 (2017-09-15)

* remove tinymce

## 2.1.0 (2017-09-14)

* views / view_partials / view_helpers cleanup
* WizardFormPartial moved from views.py to view_partials.py

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/169

## 2.0.0 (2017-09-12)

* disable WIP conditionals on review page
* update StorageHelper session api, which is now

  - StorageHelper.current_data_from_storage
  - StorageHelper.add_data_to_storage
  - StorageHelper.init_storage

        https://github.com/SexualHealthInnovations/django-wizard-builder/pull/164

## 1.0.6 (2017-09-06)

* show unanswered questions on the review page

## 1.0.2 (2017-08-22)

* add debug logging to wizard manager

## 1.0.0 (2017-08-20)

* first major release version
* lock api to callisto-core

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/125

## 0.5.5 (2017-08-18)

* Refactor wizard views into a FormView
* remove all django-formtools legacy code
* added fixtures and template defaults
* added some heroku setup management commands
* include tests in release

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/118
    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/120
    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/125

## 0.3.2 (2017-08-07)

* add wizard form templates
* remove SingleLineTextWithMap, MultiLineText, Date
* add Choice extra widgets
* add django-widget-tweaks dependency

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/112

## 0.3.1 (2017-08-04)

* pypi cleanup release

## 0.3.0 (2017-08-03)

* remove models.PageBase, TextPage, Conditional
* rename models.QuestionPage to models.Page
* remove formtools dependency
* add django-tinymce4-lite dependency

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/95

## 0.2.1 (2017-06-12)

* change models.PageBase.site to models.PageBase.sites

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/72

## 0.1.2 (2017-05-16)

* pypi cleanup

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/49

## 0.1.1 (2017-05-16)

* add dumpdata downcasting disable hack

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/44

## 0.1.0 (2017-05-16)

* remove django-polymorphic, and replace it with django-model-utils

    https://github.com/SexualHealthInnovations/django-wizard-builder/pull/40

## 0.0.9 (2017-04-27)

* Add request domain support

## 0.0.1 (2016-05-16)

* First release on PyPI.


delivery/forms.py
from urllib.parse import parse_qs, urlsplit

import bugsnag
from django import forms
from django.conf import settings
from django.core.exceptions import ValidationError
from django.forms.formsets import formset_factory
from evaluation.models import EvalRow
from nacl.exceptions import CryptoError
from zxcvbn import password_strength

REQUIRED_ERROR = "The {0} field is required."

class NewSecretKeyForm(forms.Form):
    error_messages = {
        'key_mismatch': "The two passphrase fields didn't match.",
    }

    key = forms.CharField(max_length=64,
                           label="Your passphrase",
                           widget=forms.PasswordInput(attrs={'placeholder': 'Your passphrase',
                                                         'autocomplete': 'off'}),
                           error_messages = {'required': REQUIRED_ERROR.format("passphrase")})
    key2 = forms.CharField(max_length=64,
                           label="Repeat your passphrase",
                           widget=forms.PasswordInput(attrs={'placeholder': 'Repeat your passphrase',
                                                         'autocomplete': 'off'}),
                           error_messages = {'required': REQUIRED_ERROR.format("passphrase confirmation")})

    # from http://birdhouse.org/blog/2015/06/16/sane-password-strength-validation-for-django-with-zxcvbn/

    # Portions of the below implementation are copyright cca.edu, and are under the Educational Community License:
    # https://opensource.org/licenses/ECL-2.0

    def clean_key(self):
        # Password strength testing mostly done in JS; minimal validation here.
        password = self.cleaned_data.get('key')
        results = password_strength(password)

        if results['entropy'] < settings.PASSWORD_MINIMUM_ENTROPY:
            raise forms.ValidationError("Your password isn't strong enough.")
        return password


    def clean_key2(self):
        key1 = self.cleaned_data.get("key")
        key2 = self.cleaned_data.get("key2")
        if key1 and key2 and key1 != key2:
            raise forms.ValidationError(
                self.error_messages['key_mismatch'],
                code='key_mismatch',
            )
        return key2

class SecretKeyForm(forms.Form):
    error_messages = {
        'wrong_key': "The passphrase didn't match.",
    }

    key = forms.CharField(max_length=254,
                          label="Your passphrase",
                          widget=forms.PasswordInput(attrs={'placeholder': 'Your passphrase',
                                                        'autocomplete': 'off'}),
                          error_messages = {'required': REQUIRED_ERROR.format("passphrase")})

    def clean_key(self):
        key = self.cleaned_data.get('key')
        report = self.report
        try:
            decrypted_report = report.decrypted_report(key)
            self.decrypted_report = decrypted_report
            #save anonymous row if one wasn't saved on creation
            try:
                row = EvalRow()
                row.set_identifiers(report)
                if (EvalRow.objects.filter(record_identifier = row.record_identifier).count() == 0):
                    row.action = EvalRow.FIRST
                    row.add_report_data(decrypted_report)
                    row.save()
            except Exception as e:
                #TODO: real logging
                bugsnag.notify(e)
                pass
        except CryptoError:
            self.decrypted_report = None
            raise forms.ValidationError(
                self.error_messages['wrong_key'],
                code='wrong_key',
            )
        return key


class SubmitToSchoolForm(SecretKeyForm):
    name = forms.CharField(label="Your preferred first name:",
                                    required=False,
                                    max_length=500,
                                    widget=forms.TextInput(attrs={'placeholder': 'ex. Chris'}),)

    phone_number = forms.CharField(label="Preferred phone number to call:",
                                    required=True,
                                    max_length=50,
                                    widget=forms.TextInput(attrs={'placeholder': 'ex. (555) 555-5555'}),)

    voicemail = forms.CharField(label="Is it ok to leave a voicemail? If so, what would you like the message to refer to?",
                                    required=False,
                                    max_length=256,
                                    widget=forms.TextInput(attrs={'placeholder': "ex. Yes, please just say you're following up from Callisto."}),)

    email = forms.EmailField(label="If you can't be reached by phone, what's the best email address to reach you?",
                                    required=True,
                                    max_length=256,
                                    widget=forms.TextInput(attrs={'placeholder': 'ex. myname@gmail.com'}),)

    contact_notes = forms.CharField(label="Any notes on what time of day is best to reach you? A {0} staff member will "
                                          "try their best to accommodate your needs.".format(settings.SCHOOL_SHORTNAME),
                                    required=False,
                                    widget=forms.Textarea(attrs={'placeholder': "ex. I have class from 9-2 most weekdays and work in the evenings, so mid afternoon is best (around 2:30-5pm)."
                                                                                "Wednesdays and Thursdays I have class the entire day, so those days don't work at all.",
                                                                 'max_length': 5000}),)

    email_confirmation = forms.ChoiceField(choices = [(True, "Yes"), (False, "No, thanks")],
                                           label="Would you like us to send you a confirmation email with information about your rights in the reporting process, and where to get support and find resources on campus?",
                                           required=True,
                                           widget = forms.RadioSelect)

    def __init__(self, user, report, *args, **kwargs):
        super(SecretKeyForm, self).__init__(*args, **kwargs)
        self.user = user
        self.report = report
        self.fields['key'].widget.attrs['placeholder'] = 'ex. I am a muffin baking ninja'

        #TODO: populate these intelligently if we have the data elsewhere (in match reports, other reports)
        self.fields['name'].initial = report.contact_name
        self.fields['phone_number'].initial = report.contact_phone
        self.fields['voicemail'].initial = report.contact_voicemail
        self.fields['email'].initial = report.contact_email or user.email
        self.fields['contact_notes'].initial = report.contact_notes

class SubmitToMatchingForm(forms.Form):
    perp_name = forms.CharField(label="Perpetrator's Name",
                            required=False,
                            max_length=500,
                            widget=forms.TextInput(attrs={'placeholder': 'ex. John Jacob Jingleheimer Schmidt',}),)

    perp = forms.URLField(label="Perpetrator's Facebook URL",
                            required=True,
                            max_length=500,
                            widget=forms.TextInput(attrs={'placeholder': 'ex. http://www.facebook.com/johnsmithfakename'}),)

    def clean_perp(self):
        raw_url = self.cleaned_data.get('perp')
        url_parts = urlsplit(raw_url)
        #check if Facebook
        domain = url_parts[1]
        if not (domain=='facebook.com' or domain.endswith('.facebook.com')):
            raise ValidationError('Please enter a valid Facebook profile URL.', code='notfacebook')
        path = url_parts[2].strip('/').split('/')[0]
        generic_fb_urls = ['messages', 'hashtag', 'events', 'pages', 'groups', 'bookmarks', 'lists', 'developers', 'topic', 'help',
                           'privacy', 'campaign', 'policies', 'support', 'settings', 'games']
        if path == "profile.php":
            path = parse_qs(url_parts[3]).get('id')[0]
        if not path or path == "" or path.endswith('.php') or path in generic_fb_urls:
            raise ValidationError('Please enter a valid Facebook profile URL.', code='notfacebook')
        else:
            return path

SubmitToMatchingFormSet = formset_factory(SubmitToMatchingForm, extra=0, min_num=1)
